pkg/humio/oauth.go

```diff
@@ -0,0 +1,103 @@
+package humio
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/log"
+)
+
+type OAuth2TokenResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int    `json:"expires_in"`
+}
+
+func (c *Client) fetchOAuth2Token() error {
```

Comment:
In general, if the auth server isn’t rate-limited, this function is probably fine. Otherwise, using `golang.org/x/sync/singleflight` might be helpful here so that only one goroutine does the token refresh and others either block or reuse the result.

Reply:
oooo great suggestion!